go/store/prolly/message/vector_index.go

```diff
@@ -0,0 +1,215 @@
+	}
+	serial.VectorIndexNodeAddTreeLevel(b, uint8(level))
+
+	return serial.FinishMessage(b, serial.VectorIndexNodeEnd(b), vectorIvfFileID)
+}
+
+func getVectorIndexKeysAndValues(msg serial.Message) (keys, values ItemAccess, level, count uint16, err error) {
+	var pm serial.VectorIndexNode
+	err = serial.InitVectorIndexNodeRoot(&pm, msg, serial.MessagePrefixSz)
+	if err != nil {
+		return
+	}
+	keys.bufStart = lookupVectorOffset(vectorIvfKeyItemBytesVOffset, pm.Table())
+	keys.bufLen = uint16(pm.KeyItemsLength())
+	keys.offStart = lookupVectorOffset(vectorIvfKeyOffsetsVOffset, pm.Table())
+	keys.offLen = uint16(pm.KeyOffsetsLength() * uint16Size)
+
+	count = (keys.offLen / 2) - 1
+	level = uint16(pm.TreeLevel())
+
+	vv := pm.ValueItemsBytes()
+	if vv != nil {
+		values.bufStart = lookupVectorOffset(vectorIvfValueItemBytesVOffset, pm.Table())
+		values.bufLen = uint16(pm.ValueItemsLength())
+		values.offStart = lookupVectorOffset(vectorIvfValueOffsetsVOffset, pm.Table())
+		values.offLen = uint16(pm.ValueOffsetsLength() * uint16Size)
+	} else {
+		values.bufStart = lookupVectorOffset(vectorIvfAddressArrayBytesVOffset, pm.Table())
+		values.bufLen = uint16(pm.AddressArrayLength())
+		values.itemWidth = hash.ByteLen
+	}
+	return
+}
+
+func walkVectorIndexAddresses(ctx context.Context, msg serial.Message, cb func(ctx context.Context, addr hash.Hash) error) error {
+	var pm serial.VectorIndexNode
+	err := serial.InitVectorIndexNodeRoot(&pm, msg, serial.MessagePrefixSz)
+	if err != nil {
+		return err
+	}
+	arr := pm.AddressArrayBytes()
+	for i := 0; i < len(arr)/hash.ByteLen; i++ {
+		addr := hash.New(arr[i*addrSize : (i+1)*addrSize])
+		if err := cb(ctx, addr); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func getVectorIndexCount(msg serial.Message) (uint16, error) {
+	var pm serial.VectorIndexNode
+	err := serial.InitVectorIndexNodeRoot(&pm, msg, serial.MessagePrefixSz)
+	if err != nil {
+		return 0, err
+	}
+	return uint16(pm.KeyOffsetsLength() - 1), nil
+}
+
+func getVectorIndexTreeLevel(msg serial.Message) (int, error) {
+	var pm serial.VectorIndexNode
+	err := serial.InitVectorIndexNodeRoot(&pm, msg, serial.MessagePrefixSz)
+	if err != nil {
+		return 0, fb.ErrTableHasUnknownFields
+	}
+	return int(pm.TreeLevel()), nil
+}
+
+func getVectorIndexTreeCount(msg serial.Message) (int, error) {
+	var pm serial.VectorIndexNode
+	err := serial.InitVectorIndexNodeRoot(&pm, msg, serial.MessagePrefixSz)
+	if err != nil {
+		return 0, fb.ErrTableHasUnknownFields
+	}
+	return int(pm.TreeCount()), nil
+}
+
+func getVectorIndexSubtrees(msg serial.Message) ([]uint64, error) {
+	sz, err := getVectorIndexCount(msg)
+	if err != nil {
+		return nil, err
+	}
+
+	var pm serial.VectorIndexNode
+	n := fb.GetUOffsetT(msg[serial.MessagePrefixSz:])
+	err = pm.Init(msg, serial.MessagePrefixSz+n)
+	if err != nil {
+		return nil, err
+	}
+	if serial.ProllyTreeNodeNumFields < pm.Table().NumFields() {
+		return nil, fb.ErrTableHasUnknownFields
+	}
```

Comment:
Handled by gen/fb/serial/vectorindexnode.go, no?

Reply:
Looks like you're right. I was copying the code from `prolly_map.go`, but this seems unnecessary.